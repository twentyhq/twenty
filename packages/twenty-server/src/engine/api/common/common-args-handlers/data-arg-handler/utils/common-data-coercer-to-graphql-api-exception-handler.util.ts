import { assertUnreachable } from 'twenty-shared/utils';

import {
  type CommonDataCoercerException,
  CommonDataCoercerExceptionCode,
} from 'src/engine/api/common/common-args-handlers/data-arg-handler/errors/common-data-coercer.exception';
import { UserInputError } from 'src/engine/core-modules/graphql/utils/graphql-errors.util';

export const commonDataCoercerToGraphqlApiExceptionHandler = (
  error: CommonDataCoercerException,
) => {
  switch (error.code) {
    case CommonDataCoercerExceptionCode.INVALID_COMPOSITE_FIELD:
    case CommonDataCoercerExceptionCode.INVALID_NUMBER:
    case CommonDataCoercerExceptionCode.INVALID_TEXT:
    case CommonDataCoercerExceptionCode.INVALID_DATE_OR_DATE_TIME:
    case CommonDataCoercerExceptionCode.INVALID_BOOLEAN:
    case CommonDataCoercerExceptionCode.INVALID_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_MULTI_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_UUID:
    case CommonDataCoercerExceptionCode.INVALID_ARRAY:
    case CommonDataCoercerExceptionCode.INVALID_RAW_JSON:
      throw new UserInputError(error);
    default: {
      return assertUnreachable(error.code);
    }
  }
};
