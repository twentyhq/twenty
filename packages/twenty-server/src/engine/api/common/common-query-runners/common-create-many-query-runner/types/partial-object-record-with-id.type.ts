import { type ObjectRecord } from 'twenty-shared/types';

export type PartialObjectRecordWithId = Partial<ObjectRecord> & { id: string };
