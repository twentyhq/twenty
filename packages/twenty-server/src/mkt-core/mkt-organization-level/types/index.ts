export * from './organization-level.types';
export * from './permissions.type';
