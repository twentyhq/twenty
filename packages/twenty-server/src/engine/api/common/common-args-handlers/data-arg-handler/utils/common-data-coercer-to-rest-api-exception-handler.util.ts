import { BadRequestException } from '@nestjs/common';

import { assertUnreachable } from 'twenty-shared/utils';

import {
  type CommonDataCoercerException,
  CommonDataCoercerExceptionCode,
} from 'src/engine/api/common/common-args-handlers/data-arg-handler/errors/common-data-coercer.exception';

export const commonDataCoercerToRestApiExceptionHandler = (
  error: CommonDataCoercerException,
) => {
  switch (error.code) {
    case CommonDataCoercerExceptionCode.INVALID_COMPOSITE_FIELD:
    case CommonDataCoercerExceptionCode.INVALID_NUMBER:
    case CommonDataCoercerExceptionCode.INVALID_TEXT:
    case CommonDataCoercerExceptionCode.INVALID_DATE_OR_DATE_TIME:
    case CommonDataCoercerExceptionCode.INVALID_BOOLEAN:
    case CommonDataCoercerExceptionCode.INVALID_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_MULTI_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_UUID:
    case CommonDataCoercerExceptionCode.INVALID_ARRAY:
    case CommonDataCoercerExceptionCode.INVALID_RAW_JSON:
      throw new BadRequestException(error.message);
    default: {
      return assertUnreachable(error.code);
    }
  }
};
