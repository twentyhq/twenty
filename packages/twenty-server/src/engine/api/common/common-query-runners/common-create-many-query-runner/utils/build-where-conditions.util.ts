import { isDefined } from 'twenty-shared/utils';
import { type FindOperator, In } from 'typeorm';

import { type ObjectRecord } from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

import { getValueFromPath } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/utils/get-value-from-path.util';

export const buildWhereConditions = (
  records: Partial<ObjectRecord>[],
  conflictingFields: {
    baseField: string;
    fullPath: string;
    column: string;
  }[],
): Record<string, FindOperator<string>>[] => {
  const whereConditions = [];

  for (const field of conflictingFields) {
    const fieldValues = records
      .map((record) => getValueFromPath(record, field.fullPath))
      .filter(isDefined);

    if (fieldValues.length > 0) {
      whereConditions.push({ [field.column]: In(fieldValues) });
    }
  }

  return whereConditions;
};
