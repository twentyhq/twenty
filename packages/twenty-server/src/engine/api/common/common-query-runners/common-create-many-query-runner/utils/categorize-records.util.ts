import { isDefined } from 'twenty-shared/utils';

import { type ObjectRecord } from 'twenty-shared/types';

import { type PartialObjectRecordWithId } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/types/partial-object-record-with-id.type';
import { getMatchingRecordId } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/utils/get-matching-record-id.util';

export const categorizeRecords = (
  records: Partial<ObjectRecord>[],
  conflictingFields: {
    baseField: string;
    fullPath: string;
    column: string;
  }[],
  existingRecords: PartialObjectRecordWithId[],
): {
  recordsToUpdate: PartialObjectRecordWithId[];
  recordsToInsert: Partial<ObjectRecord>[];
} => {
  const recordsToUpdate: PartialObjectRecordWithId[] = [];
  const recordsToInsert: Partial<ObjectRecord>[] = [];

  for (const record of records) {
    const matchingRecordId = getMatchingRecordId(
      record,
      conflictingFields,
      existingRecords,
    );

    if (isDefined(matchingRecordId)) {
      recordsToUpdate.push({ ...record, id: matchingRecordId });
    } else {
      recordsToInsert.push(record);
    }
  }

  return { recordsToUpdate, recordsToInsert };
};
