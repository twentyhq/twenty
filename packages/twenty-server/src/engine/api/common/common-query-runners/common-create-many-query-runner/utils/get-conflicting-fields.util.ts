import { capitalize } from 'twenty-shared/utils';

import { compositeTypeDefinitions } from 'src/engine/metadata-modules/field-metadata/composite-types';
import { type ObjectMetadataItemWithFieldMaps } from 'src/engine/metadata-modules/types/object-metadata-item-with-field-maps';

export const getConflictingFields = (
  objectMetadataItemWithFieldMaps: ObjectMetadataItemWithFieldMaps,
): {
  baseField: string;
  fullPath: string;
  column: string;
}[] => {
  return Object.values(objectMetadataItemWithFieldMaps.fieldsById)
    .filter((field) => field.isUnique || field.name === 'id')
    .flatMap((field) => {
      const compositeType = compositeTypeDefinitions.get(field.type);

      if (!compositeType) {
        return [
          {
            baseField: field.name,
            fullPath: field.name,
            column: field.name,
          },
        ];
      }

      const property = compositeType.properties.find(
        (prop) => prop.isIncludedInUniqueConstraint,
      );

      return property
        ? [
            {
              baseField: field.name,
              fullPath: `${field.name}.${property.name}`,
              column: `${field.name}${capitalize(property.name)}`,
            },
          ]
        : [];
    });
};
