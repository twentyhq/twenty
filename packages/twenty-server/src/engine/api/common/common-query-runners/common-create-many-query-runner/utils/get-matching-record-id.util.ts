import { msg } from '@lingui/core/macro';
import { type ObjectRecord } from 'twenty-shared/types';
import { isDefined } from 'twenty-shared/utils';

import { type PartialObjectRecordWithId } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/types/partial-object-record-with-id.type';
import { getValueFromPath } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/utils/get-value-from-path.util';
import {
  CommonQueryRunnerException,
  CommonQueryRunnerExceptionCode,
} from 'src/engine/api/common/common-query-runners/errors/common-query-runner.exception';

export const getMatchingRecordId = (
  record: Partial<ObjectRecord>,
  conflictingFields: {
    baseField: string;
    fullPath: string;
    column: string;
  }[],
  existingRecords: PartialObjectRecordWithId[],
): string | undefined => {
  const matchingRecordIds = conflictingFields.reduce<string[]>((acc, field) => {
    const requestFieldValue = getValueFromPath(record, field.fullPath);

    const matchingRecord = existingRecords.find((existingRecord) => {
      const existingFieldValue = getValueFromPath(
        existingRecord,
        field.fullPath,
      );

      return (
        isDefined(existingFieldValue) &&
        existingFieldValue === requestFieldValue
      );
    });

    if (isDefined(matchingRecord)) {
      acc.push(matchingRecord.id);
    }

    return acc;
  }, []);

  if ([...new Set(matchingRecordIds)].length > 1) {
    const conflictingFieldsValues = conflictingFields
      .map((field) => {
        const value = getValueFromPath(record, field.fullPath);

        return isDefined(value) ? `${field.fullPath}: ${value}` : undefined;
      })
      .filter(isDefined)
      .join(', ');

    throw new CommonQueryRunnerException(
      `Multiple records found with the same unique field values for ${conflictingFieldsValues}. Cannot determine which record to update.`,
      CommonQueryRunnerExceptionCode.UPSERT_MULTIPLE_MATCHING_RECORDS_CONFLICT,
      {
        userFriendlyMessage: msg`Multiple records found with the same unique field values for ${conflictingFieldsValues}. Cannot determine which record to update.`,
      },
    );
  }

  return matchingRecordIds[0];
};
