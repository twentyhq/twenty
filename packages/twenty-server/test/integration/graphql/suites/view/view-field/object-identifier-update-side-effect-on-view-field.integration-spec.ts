import { FlatViewField } from 'src/engine/core-modules/view/flat-view/types/flat-view-field.type';
import {
  type ViewFieldTestSetup,
  cleanupViewFieldTestV2,
  setupViewFieldTestV2,
} from 'test/integration/graphql/suites/view/utils/setup-view-field-test-v2.util';
import { createOneFieldMetadata } from 'test/integration/metadata/suites/field-metadata/utils/create-one-field-metadata.util';
import { updateOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/update-one-object-metadata.util';
import { createOneCoreViewField } from 'test/integration/metadata/suites/view-field/utils/create-one-core-view-field.util';
import { deleteOneCoreViewField } from 'test/integration/metadata/suites/view-field/utils/delete-one-core-view-field.util';
import { destroyOneCoreViewField } from 'test/integration/metadata/suites/view-field/utils/destroy-one-core-view-field.util';
import { findCoreViewFields } from 'test/integration/metadata/suites/view-field/utils/find-core-view-fields.util';
import { updateOneCoreViewField } from 'test/integration/metadata/suites/view-field/utils/update-one-core-view-field.util';
import { extractRecordIdsAndDatesAsExpectAny } from 'test/utils/extract-record-ids-and-dates-as-expect-any';
import { FieldMetadataType } from 'twenty-shared/types';

describe('View Field Resolver - Successful object metadata identifier update side effect on view field', () => {
  let testSetup: ViewFieldTestSetup & {
    testLabelIdentifierViewFieldId: string;
  };
  beforeAll(async () => {
    const { testFieldMetadataId, testObjectMetadataId, testViewId } =
      await setupViewFieldTestV2();

    await updateOneObjectMetadata({
      input: {
        idToUpdate: testObjectMetadataId,
        updatePayload: {
          labelIdentifierFieldMetadataId: testFieldMetadataId,
        },
      },
      expectToFail: false,
    });

    const {
      data: { getCoreViewFields },
    } = await findCoreViewFields({
      viewId: testViewId,
      expectToFail: false,
      gqlFields: `
        id
        fieldMetadataId
        viewId
      `,
    });
    const testLabelIdentifierViewFieldId = getCoreViewFields[0].id;
    testSetup = {
      testFieldMetadataId,
      testObjectMetadataId,
      testViewId,
      testLabelIdentifierViewFieldId,
    };
  });

  afterAll(async () => {
    await cleanupViewFieldTestV2(testSetup.testObjectMetadataId);
  });

  it('should create a view field on label identifier object metadata update if it does not exist on view', async () => {
    const { testFieldMetadataId, testViewId } = testSetup;
    const {
      data: { getCoreViewFields },
    } = await findCoreViewFields({
      viewId: testViewId,
      expectToFail: false,
      gqlFields: `
        id
        fieldMetadataId
        viewId
      `,
    });

    expect(getCoreViewFields).toMatchObject<
      Pick<FlatViewField, 'id' | 'fieldMetadataId' | 'viewId'>[]
    >([
      {
        id: testSetup.testLabelIdentifierViewFieldId,
        fieldMetadataId: testFieldMetadataId,
        viewId: testViewId,
      },
    ]);
  });

  it('Should not allow deleting a label identifier view field', async () => {
    const { errors } = await deleteOneCoreViewField({
      input: { id: testSetup.testLabelIdentifierViewFieldId },
      expectToFail: true,
    });

    // TODO create a util to expect a core engine v2 error
    expect(errors.length).toBe(1);
    const [firstError] = errors;
    expect(firstError.extensions.code).not.toBe('INTERNAL_SERVER_ERROR');
    expect(firstError).toMatchSnapshot(
      extractRecordIdsAndDatesAsExpectAny(firstError),
    );
  });

  it('Should not allow destroying a label identifier view field', async () => {
    const { errors } = await destroyOneCoreViewField({
      input: { id: testSetup.testLabelIdentifierViewFieldId },
      expectToFail: true,
    });

    expect(errors.length).toBe(1);
    const [firstError] = errors;
    expect(firstError.extensions.code).not.toBe('INTERNAL_SERVER_ERROR');
    expect(firstError).toMatchSnapshot(
      extractRecordIdsAndDatesAsExpectAny(firstError),
    );
  });

  it('Should not allow updating a label identifier view field visibility to false', async () => {
    const { errors } = await updateOneCoreViewField({
      input: {
        id: testSetup.testLabelIdentifierViewFieldId,
        update: { isVisible: false },
      },
      expectToFail: true,
    });

    expect(errors.length).toBe(1);
    const [firstError] = errors;
    expect(firstError.extensions.code).not.toBe('INTERNAL_SERVER_ERROR');
    expect(firstError).toMatchSnapshot(
      extractRecordIdsAndDatesAsExpectAny(firstError),
    );
  });

  it('Should not allow creating a view field with a position lower than the label idenfitier view field', async () => {
    const {
      data: {
        createOneField: { id: fieldMetadataId },
      },
    } = await createOneFieldMetadata({
      expectToFail: false,
      input: {
        name: 'nonLabelIdentifierField',
        label: 'Test Field',
        type: FieldMetadataType.TEXT,
        objectMetadataId: testSetup.testObjectMetadataId,
        isLabelSyncedWithName: false,
      },
      gqlFields: `
          id
          name
          label
          isLabelSyncedWithName
        `,
    });

    const { errors } = await createOneCoreViewField({
      input: {
        fieldMetadataId: fieldMetadataId,
        viewId: testSetup.testViewId,
        position: -42,
      },
      expectToFail: true,
    });

    expect(errors.length).toBe(1);
    const [firstError] = errors;
    expect(firstError.extensions.code).not.toBe('INTERNAL_SERVER_ERROR');
    expect(firstError).toMatchSnapshot(
      extractRecordIdsAndDatesAsExpectAny(firstError),
    );
  });

  it.only('Should not allow updated labelIdentifier view field with a position higher than existing other view field', async () => {
    const {
      data: {
        createOneField: { id: fieldMetadataId },
      },
    } = await createOneFieldMetadata({
      expectToFail: false,
      input: {
        name: 'nonLabelIdentifierField',
        label: 'Test Field',
        type: FieldMetadataType.TEXT,
        objectMetadataId: testSetup.testObjectMetadataId,
        isLabelSyncedWithName: false,
      },
      gqlFields: `
          id
          name
          label
          isLabelSyncedWithName
        `,
    });

    await createOneCoreViewField({
      input: {
        fieldMetadataId: fieldMetadataId,
        viewId: testSetup.testViewId,
        position: 42,
      },
      expectToFail: false,
    });

    const { errors } = await updateOneCoreViewField({
      input: {
        id: testSetup.testLabelIdentifierViewFieldId,
        update: {
          position: 43,
        },
      },
      expectToFail: true,
    });
    expect(errors.length).toBe(1);
    const [firstError] = errors;
    expect(firstError.extensions.code).not.toBe('INTERNAL_SERVER_ERROR');
    expect(firstError).toMatchSnapshot(
      extractRecordIdsAndDatesAsExpectAny(firstError),
    );
  });
});
