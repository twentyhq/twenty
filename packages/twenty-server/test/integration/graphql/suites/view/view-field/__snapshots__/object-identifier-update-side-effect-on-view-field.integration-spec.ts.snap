// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Field Resolver - Successful object metadata identifier update side effect on view field Should not allow creating a view field with a position lower than the label identifier view field 1`] = `
{
  "extensions": {
    "code": "BAD_USER_INPUT",
    "errors": {
      "cronTrigger": [],
      "databaseEventTrigger": [],
      "fieldMetadata": [],
      "index": [],
      "objectMetadata": [],
      "routeTrigger": [],
      "serverlessFunction": [],
      "view": [],
      "viewField": [
        {
          "errors": [
            {
              "code": "INVALID_VIEW_DATA",
              "message": "View field position cannot be lower than label identifier view field position",
              "userFriendlyMessage": "View field position cannot be lower than label identifier view field position",
            },
          ],
          "flatEntityMinimalInformation": {
            "fieldMetadataId": Any<String>,
            "id": Any<String>,
            "viewId": Any<String>,
          },
          "status": "fail",
          "type": "create_view_field",
        },
      ],
      "viewFilter": [],
      "viewGroup": [],
    },
    "message": "Validation failed for 0 object(s) and 0 field(s)",
    "summary": {
      "invalidCronTrigger": 0,
      "invalidDatabaseEventTrigger": 0,
      "invalidFieldMetadata": 0,
      "invalidIndex": 0,
      "invalidObjectMetadata": 0,
      "invalidRouteTrigger": 0,
      "invalidServerlessFunction": 0,
      "invalidView": 0,
      "invalidViewField": 0,
      "invalidViewFilter": 0,
      "invalidViewGroup": 0,
      "totalErrors": 0,
    },
    "userFriendlyMessage": "Validation failed for 0 object(s) and 0 field(s)",
  },
  "message": "Multiple validation errors occurred while creating view field",
  "name": "GraphQLError",
}
`;

exports[`View Field Resolver - Successful object metadata identifier update side effect on view field Should not allow deleting a label identifier view field 1`] = `
{
  "extensions": {
    "code": "BAD_USER_INPUT",
    "errors": {
      "cronTrigger": [],
      "databaseEventTrigger": [],
      "fieldMetadata": [],
      "index": [],
      "objectMetadata": [],
      "routeTrigger": [],
      "serverlessFunction": [],
      "view": [],
      "viewField": [
        {
          "errors": [
            {
              "code": "INVALID_VIEW_DATA",
              "message": "Label identifier view field cannot be deleted",
              "userFriendlyMessage": "Label identifier view field cannot be deleted",
            },
          ],
          "flatEntityMinimalInformation": {
            "fieldMetadataId": Any<String>,
            "id": Any<String>,
            "viewId": Any<String>,
          },
          "status": "fail",
          "type": "update_view_field",
        },
      ],
      "viewFilter": [],
      "viewGroup": [],
    },
    "message": "Validation failed for 0 object(s) and 0 field(s)",
    "summary": {
      "invalidCronTrigger": 0,
      "invalidDatabaseEventTrigger": 0,
      "invalidFieldMetadata": 0,
      "invalidIndex": 0,
      "invalidObjectMetadata": 0,
      "invalidRouteTrigger": 0,
      "invalidServerlessFunction": 0,
      "invalidView": 0,
      "invalidViewField": 0,
      "invalidViewFilter": 0,
      "invalidViewGroup": 0,
      "totalErrors": 0,
    },
    "userFriendlyMessage": "Validation failed for 0 object(s) and 0 field(s)",
  },
  "message": "Multiple validation errors occurred while deleting view field",
  "name": "GraphQLError",
}
`;

exports[`View Field Resolver - Successful object metadata identifier update side effect on view field Should not allow destroying a label identifier view field 1`] = `
{
  "extensions": {
    "code": "BAD_USER_INPUT",
    "errors": {
      "cronTrigger": [],
      "databaseEventTrigger": [],
      "fieldMetadata": [],
      "index": [],
      "objectMetadata": [],
      "routeTrigger": [],
      "serverlessFunction": [],
      "view": [],
      "viewField": [
        {
          "errors": [
            {
              "code": "INVALID_VIEW_DATA",
              "message": "Label identifier view field cannot be deleted",
              "userFriendlyMessage": "Label identifier view field cannot be deleted",
            },
          ],
          "flatEntityMinimalInformation": {
            "id": Any<String>,
          },
          "status": "fail",
          "type": "delete_view_field",
        },
      ],
      "viewFilter": [],
      "viewGroup": [],
    },
    "message": "Validation failed for 0 object(s) and 0 field(s)",
    "summary": {
      "invalidCronTrigger": 0,
      "invalidDatabaseEventTrigger": 0,
      "invalidFieldMetadata": 0,
      "invalidIndex": 0,
      "invalidObjectMetadata": 0,
      "invalidRouteTrigger": 0,
      "invalidServerlessFunction": 0,
      "invalidView": 0,
      "invalidViewField": 0,
      "invalidViewFilter": 0,
      "invalidViewGroup": 0,
      "totalErrors": 0,
    },
    "userFriendlyMessage": "Validation failed for 0 object(s) and 0 field(s)",
  },
  "message": "Multiple validation errors occurred while deleting view field",
  "name": "GraphQLError",
}
`;

exports[`View Field Resolver - Successful object metadata identifier update side effect on view field Should not allow updating a label identifier view field visibility to false 1`] = `
{
  "extensions": {
    "code": "BAD_USER_INPUT",
    "errors": {
      "cronTrigger": [],
      "databaseEventTrigger": [],
      "fieldMetadata": [],
      "index": [],
      "objectMetadata": [],
      "routeTrigger": [],
      "serverlessFunction": [],
      "view": [],
      "viewField": [
        {
          "errors": [
            {
              "code": "INVALID_VIEW_DATA",
              "message": "Label identifier view field has to be visible",
              "userFriendlyMessage": "Label identifier view field has to be visible",
            },
          ],
          "flatEntityMinimalInformation": {
            "fieldMetadataId": Any<String>,
            "id": Any<String>,
            "viewId": Any<String>,
          },
          "status": "fail",
          "type": "update_view_field",
        },
      ],
      "viewFilter": [],
      "viewGroup": [],
    },
    "message": "Validation failed for 0 object(s) and 0 field(s)",
    "summary": {
      "invalidCronTrigger": 0,
      "invalidDatabaseEventTrigger": 0,
      "invalidFieldMetadata": 0,
      "invalidIndex": 0,
      "invalidObjectMetadata": 0,
      "invalidRouteTrigger": 0,
      "invalidServerlessFunction": 0,
      "invalidView": 0,
      "invalidViewField": 0,
      "invalidViewFilter": 0,
      "invalidViewGroup": 0,
      "totalErrors": 0,
    },
    "userFriendlyMessage": "Validation failed for 0 object(s) and 0 field(s)",
  },
  "message": "Multiple validation errors occurred while updating view field",
  "name": "GraphQLError",
}
`;

exports[`View Field Resolver - Successful object metadata identifier update side effect on view field Should not allow updating labelIdentifier view field with a position higher than existing other view field 1`] = `
{
  "extensions": {
    "code": "BAD_USER_INPUT",
    "errors": {
      "cronTrigger": [],
      "databaseEventTrigger": [],
      "fieldMetadata": [],
      "index": [],
      "objectMetadata": [],
      "routeTrigger": [],
      "serverlessFunction": [],
      "view": [],
      "viewField": [
        {
          "errors": [
            {
              "code": "INVALID_VIEW_DATA",
              "message": "Label identifier view field has to be in the lowest position",
              "userFriendlyMessage": "Label identifier view field has to be in the lowest position",
            },
          ],
          "flatEntityMinimalInformation": {
            "fieldMetadataId": Any<String>,
            "id": Any<String>,
            "viewId": Any<String>,
          },
          "status": "fail",
          "type": "update_view_field",
        },
      ],
      "viewFilter": [],
      "viewGroup": [],
    },
    "message": "Validation failed for 0 object(s) and 0 field(s)",
    "summary": {
      "invalidCronTrigger": 0,
      "invalidDatabaseEventTrigger": 0,
      "invalidFieldMetadata": 0,
      "invalidIndex": 0,
      "invalidObjectMetadata": 0,
      "invalidRouteTrigger": 0,
      "invalidServerlessFunction": 0,
      "invalidView": 0,
      "invalidViewField": 0,
      "invalidViewFilter": 0,
      "invalidViewGroup": 0,
      "totalErrors": 0,
    },
    "userFriendlyMessage": "Validation failed for 0 object(s) and 0 field(s)",
  },
  "message": "Multiple validation errors occurred while updating view field",
  "name": "GraphQLError",
}
`;
